src/algorithms/AsymmetricForwardBackwardAdjoint.jl

```diff
@@ -0,0 +1,181 @@
+                alpha = betaR/(2*nmL);
+            elseif betaQ > temp*par*nmL
+                alpha = 2*nmL*temp/betaQ;
+            end     
+            gamma2 = 1/(betaR/2+ alpha*nmL);
+            gamma1 = 0.9/(betaQ/2+ (temp-1)*gamma2*nmL^2+ nmL/alpha); 
+        else 
+            error("this choice of theta and mu is not supported!")
+        end 
+    end
+    end
+    return AFBAIterator{I, R, T}(x0, y0, g, h, f, l, L, theta, mu, lam, betaQ, betaR, gamma1, gamma2, maxit, tol, verbose, verbose_freq, xbar, ybar, gradf, gradl, FPR_x,FPR_y)
+end
+
+################################################################################
+# Utility methods
+maxit(sol::AFBAIterator) = sol.maxit
+
+converged(sol::AFBAIterator, it) = norm(sol.FPR_x)+norm(sol.FPR_y) <= sol.tol
+
+verbose(sol::AFBAIterator) = sol.verbose > 0
+verbose(sol::AFBAIterator, it) = sol.verbose > 0 && (sol.verbose == 2 ? true : (it == 1 || it%sol.verbose_freq == 0))
+
+function display(sol::AFBAIterator)
+    @printf("%6s | %22s | %10s |\n ", "it", "gamma", "fpr")
+    @printf("------|------------------------|------------|\n")
+end
+
+function display(sol::AFBAIterator, it)
+    @printf("%6d | %7.4e, %7.4e | %7.4e |\n", it, sol.gamma1, sol.gamma2, norm(sol.FPR_x)+norm(sol.FPR_y))
+end
+
+function Base.show(io::IO, sol::AFBAIterator)
+    println(io, "Asymmetric Forward-Backward-Adjoint Splitting" )
+    if sol.theta==2
+        println(io, "theta               : $(sol.theta)" )
+    else 
+        println(io, "theta, mu           : $(sol.theta), $(sol.mu)" )
+    end
+    println(io, "fpr                 : $(norm(sol.FPR_x)+norm(sol.FPR_y))")
+    print(  io, "gamma1, gamma2      : $(sol.gamma1), $(sol.gamma2)")
+end
+
+################################################################################
+# Initialization
+
+function initialize(sol::AFBAIterator)
+    return
+end
+
+################################################################################
+# Iteration
+
+function iterate(sol::AFBAIterator{I, R,T}, it::I) where {I, R, T}
+
+    # considering when h \equiv 0
+    if typeof(sol.h)==Zero
+         # perform x-update step
+        gradient!(sol.gradf,sol.f, sol.x)
+        prox!(sol.xbar, sol.g, sol.x .-sol.gamma1*sol.gradf, sol.gamma1)
+        # perform y-update step
+        sol.FPR_x .= sol.xbar .- sol.x   # fix this later
+        sol.FPR_y .= 0   # fix this later
+        sol.x .= sol.x .+ sol.lam .*(sol.FPR_x) 
+    else
+      # perform x-update step
+        gradient!(sol.gradf,sol.f, sol.x)
+        prox!(sol.xbar, sol.g, sol.x .- sol.L'*(sol.gamma1*sol.y).-sol.gamma1*sol.gradf, sol.gamma1)
+        # perform y-update step
+        gradient!(sol.gradl,sol.l, sol.y)   ############fix  # l is basically l^*!! hence Zero() by default 
+        prox!(sol.ybar, Conjugate(sol.h), sol.y .+ sol.L*(sol.gamma2*(sol.theta*sol.xbar.+(1-sol.theta)*sol.x)).-sol.gamma2*sol.gradl, sol.gamma2) 
```

Comment:
Can we avoid introducing two temporary variables for each prox? For example at line 160, is it ok to use the already allocated sol.xbar instead of introducing temp_x?  
 *  temp_y  = - sol.gamma1*sol.y
 *  Ac_mul_B!(sol.xbar, sol.L,temp_y)
*  sol.xbar .+=  sol.x .- sol.gamma1*sol.gradf
*  prox!(sol.xbar, sol.g, sol.xbar, sol.gamma1)

Reply:
I see that sol.xbar being both input and output for Prox! is not working. It is not quite clear to me why. I updated the code. What do you think? 

Regarding hconj, lconj, The issue is the special cases. Since Conjugate doesn't work for Zero().  
Another issue is the function l(x) in the infimal convolution h \oblong l (x). The algorithm requires \nabla l^*(y). At the moment the user should input l^* instead of l.

Reply:
So what you mean is to preallocate temp_x, temp_y globally? 

Another question: Is there any difference between  
> sol.temp_y .-= sol.gradl
sol.temp_y .*= sol.gamma2

and 
sol.temp_y .-= sol.gamma2 .* sol.gradl

Reply:
yes sorry wrong example. I meant to say is there a difference between 
> sol.temp_x .= sol.theta.*sol.xbar .+ (1-sol.theta).*sol.x
A_mul_B!(sol.temp_y, sol.L, sol.temp_x)
sol.temp_y .-= sol.gradl
sol.temp_y . *= sol.gamma2
sol.temp_y .+= sol.y

and 

> temp_x .= sol.gamma2 * ( (sol.theta .* sol.xbar) .+ ((1-sol.theta) * sol.x) )
A_mul_B!(sol.temp_y, sol.L, sol.temp_x)
temp_y .+=  sol.y .- sol.gamma2*sol.gradl

apart from the pre-multiplication by gamma2?